File: LinkedList/DoublyLinkList/DLL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DLL:
    def __init__(self):
        self.head = None

    def Insert(self):
        new_data = int(input("\nEnter Data >> "))
        print("\n1. Insert at start\n2. Middle\n3. End")
        ch1 = int(input(">> "))
        new_node = Node(new_data)

        if ch1 == 1:
            new_node.next = self.head
            if self.head is not None:
                self.head.prev = new_node
            self.head = new_node
        elif ch1 == 2:
            print("\nAfter which Node you want to insert data >> ")
            ch1 = int(input())
            temp = self.head
            while temp is not None and temp.data != ch1:
                temp = temp.next
            if temp is None:
                print("\nNode not present")
                return
            new_node.next = temp.next
            new_node.prev = temp
            if temp.next is not None:
                temp.next.prev = new_node
            temp.next = new_node
        elif ch1 == 3:
            new_node.next = None
            if self.head is None:
                new_node.prev = None
                self.head = new_node
            else:
                temp = self.head
                while temp.next is not None:
                    temp = temp.next
                temp.next = new_node
                new_node.prev = temp

    def Search(self):
        if self.head is None:
            print("\nThe DLL is empty. Cannot perform search.")
            return

        target_data = int(input("\nEnter the data to search: "))
        temp = self.head
        while temp is not None:
            if temp.data == target_data:
                print("\nData found in the DLL.")
                return
            temp = temp.next

        print("\nData not found in the DLL.")

    def Delete(self):
        if self.head is None:
            print("\nThe DLL is empty. Cannot perform delete.")
            return

        target_data = int(input("\nEnter the data to delete: "))
        temp = self.head
        while temp is not None and temp.data != target_data:
            temp = temp.next

        if temp is None:
            print("\nData not found in the DLL. Cannot delete.")
            return

        if temp.prev is not None:
            temp.prev.next = temp.next
        else:
            self.head = temp.next

        if temp.next is not None:
            temp.next.prev = temp.prev

        print("\nData deleted from the DLL.")

    def Update(self):
        if self.head is None:
            print("\nThe DLL is empty. Cannot perform update.")
            return

        target_data = int(input("\nEnter the data to update: "))
        temp = self.head
        while temp is not None and temp.data != target_data:
            temp = temp.next

        if temp is None:
            print("\nData not found in the DLL. Cannot update.")
            return

        new_data = int(input("\nEnter the new data value: "))
        temp.data = new_data
        print("\nData updated in the DLL.")

    def Display(self):
        if self.head is None:
            print("\nThe DLL is empty")
            return

        temp = self.head
        print("NULL <-> ", end="")
        while temp is not None:
            print(f"{temp.data} <-> ", end="")
            temp = temp.next
        print("NULL")

    def Count(self):
        count = 0
        temp = self.head
        while temp is not None:
            count += 1
            temp = temp.next

        print(f"\nNumber of nodes in the DLL: {count}")

    def Reverse(self):
        current = self.head
        temp = None

        while current is not None:
            temp = current.prev
            current.prev = current.next
            current.next = temp

            current = current.prev

        if temp is not None:
            self.head = temp.prev

        print("\nDLL reversed.")


if __name__ == "__main__":
    x = DLL()
    ch = 0
    while ch != 8:
        print("\n1. Insert\n2. Search\n3. Delete\n4. Update\n5. Display\n6. Count\n7. Reverse\n8. Exit")
        ch = int(input(">> "))
        if ch == 1:
            x.Insert()
        elif ch == 2:
            x.Search()
        elif ch == 3:
            x.Delete()
        elif ch == 4:
            x.Update()
        elif ch == 5:
            x.Display()
        elif ch == 6:
            x.Count()
        elif ch == 7:
            x.Reverse()

File: LinkedList/DoublyLinkList/DLL.java
package DSA.LinkedList.DoublyLinkList;

import java.util.Scanner;

class DLL
{
    static class Node
    {
        int data;
        Node next,prev;
        Node(int d)
        {
            data = d;
            next = null;
            prev = null;
        }
    }
    static Node head;
    int ch1 = 0, targetData = 0;
    static Scanner sc = new Scanner(System.in);

    static void Insert()
    {
        System.out.println("\nEnter Data >> ");
        int newData = sc.nextInt();
        Node newNode = new Node(newData);

        System.out.println("\n1.Insert at Start \n2.Middle \n3.End >> ");
        int ch1 = sc.nextInt();

        if (ch1 == 1) {
            newNode.next = head;
            newNode.prev = null;
            if (head != null) {
                head.prev = newNode;
            }
            head = newNode;
        } else if (ch1 == 2) {
            System.out.println("After which node you want to insert data >> ");
            ch1 = sc.nextInt();
            Node temp = head;
            while (temp != null && temp.data != ch1) {
                temp = temp.next;
            }
            if (temp == null) {
                System.out.println("\nNode not present");
                return;
            }
            newNode.next = temp.next;
            newNode.prev = temp;
            if (temp.next != null) {
                temp.next.prev = newNode;
            }
            temp.next = newNode;
        } else if (ch1 == 3) {
            newNode.next = null;
            if (head == null) {
                newNode.prev = null;
                head = newNode;
            } else {
                Node temp = head;
                while (temp.next != null) {
                    temp = temp.next;
                }
                temp.next = newNode;
                newNode.prev = temp;
            }
        }
    }

    static void Delete()
    {
        System.out.println("\nEnter the data to delete >> ");
        int targetData = sc.nextInt();
        Node temp = head;
        while (temp != null && temp.data != targetData) {
            temp = temp.next;
        }
        if (temp == null) {
            System.out.println("\nData not found in DLL");
            return;
        }
        if (temp.prev != null) {
            temp.prev.next = temp.next;
        } else {
            head = temp.next;
        }
        if (temp.next != null) {
            temp.next.prev = temp.prev;
        }
        System.out.println("\nData Deleted from the DLL");
    }

    static void Search()
    {
        System.out.println("\nEnter the data to search >> ");
        int targetData = sc.nextInt();
        int count = 0;
        Node temp = head;
        while (temp != null) {
            if (temp.data == targetData) {
                System.out.println("\nData found in the DLL at index " + count);
                return;
            }
            count++;
            temp = temp.next;
        }
        System.out.println("Data not found in the DLL");
    }

    static void Update()
    {
        System.out.println("\nEnter the data to update >> ");
        int targetData = sc.nextInt();
        Node temp = head;
        while (temp != null && temp.data != targetData) {
            temp = temp.next;
        }
        if (temp == null) {
            System.out.println("Data not found in the DLL");
            return;
        }
        System.out.println("Enter the new Data value >> ");
        temp.data = sc.nextInt();
        System.out.println("Data updated in the DLL");
    }

    static void Display()
    {
        if (head == null) {
            System.out.println("\n\nThe DLL is empty");
            return;
        }
        Node temp = head;
        System.out.print("NULL <-> ");
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        }
        System.out.print("NULL");
    }

    static void Count()
    {
        int count = 0;
        Node temp = head;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        System.out.println("\nNumber of nodes in DLL " + count);
    }

    static void Reverse()
    {
        System.out.println("\n\nDLL Before Reversed \n");
        Display();
        Node current = head;
        Node temp = null;
        while (current != null) {
            temp = current.prev;
            current.prev = current.next;
            current.next = temp;
            current = current.prev;
        }
        if (temp != null) {
            head = temp.prev;
        }
        System.out.println("\n\nDLL After Reversed \n");
        Display();
    }

    public static void main(String[] args)
    {
        DLL x = new DLL();
        int ch = 0;
        while (ch != 8) {
            System.out.print("\n1.Insert\n2.Search\n3.Delete\n4.Update\n5.Display\n6.Count\n7.Reverse\n8.Exit\n> > > ");
            ch = sc.nextInt();
            switch (ch) {
                case 1:
                    Insert();
                    break;
                case 2:
                    Search();
                    break;
                case 3:
                    Delete();
                    break;
                case 4:
                    Update();
                    break;
                case 5:
                    Display();
                    break;
                case 6:
                    Count();
                    break;
                case 7:
                    Reverse();
                    break;
                case 8:
                    break;
            }
        }
    }
}